hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java

```diff
@@ -89,8 +219,76 @@ public void put(Object key, Object value) throws CacheException {
          // any subsequent read will just see the old result with its
          // out-of-date timestamp; that result will be discarded and the
          // db query performed again.
-         putCache.put(key, value);
+         putCache.put( key, value );
       }
    }
 
-}
\ No newline at end of file
+   private class PostTransactionQueryUpdate implements Synchronization {
+      private final Transaction transaction;
+      private final Object key;
+      private volatile Object value;
+
+      public PostTransactionQueryUpdate(Transaction transaction, Object key, Object value) {
+         this.transaction = transaction;
+         this.key = key;
+         this.value = value;
+      }
+
+      public Object getValue() {
+         return value;
+      }	
+
+      public void setValue(Object value) {
+         this.value = value;
+      }
+
+      @Override
+      public void beforeCompletion() {
+      }
+
+      @Override
+      public void afterCompletion(int status) {
+         transactionContext.remove(transaction);
+         Object value = this.value;
+         if (value == null) {
+            return;
+         }
+         switch (status) {
+            case Status.STATUS_COMMITTING:
+            case Status.STATUS_COMMITTED:
+               TransactionManager tm = getTransactionManager();
+               Transaction suspended = null;
+               try {
+                  suspended = tm.suspend();
+                  if (putCacheRequiresTransaction) {
+                     tm.begin();
+                     try {
+                        putCache.put(key, value);
+                     } finally {
+                        tm.commit();
```

Comment:
You already have `org.hibernate.cache.infinispan.util.Caches` to deal with similar patterns, don't you?